orttraining/orttraining/training_api/onnxruntime_training_c_api.cc

```diff
@@ -22,6 +22,13 @@ std::vector<std::shared_ptr<onnxruntime::IExecutionProvider>> CreateProviders(
   return execution_providers;
 }
 
+static char* StrDup(const std::string& str, _Inout_ OrtAllocator* allocator) {
+  char* output_string = reinterpret_cast<char*>(allocator->Alloc(allocator, str.size() + 1));
+  memcpy(output_string, str.c_str(), str.size());
+  output_string[str.size()] = '\0';
+  return output_string;
+}
```

Comment:
Do we need the user to pass in the allocator? In this case, what is the lifetime of the char*?

I did it slightly differently in my [checkpoint property PR](https://github.com/microsoft/onnxruntime/blob/472a44f3f5732db318eb7524f43f605c8e7fc3ce/orttraining/orttraining/training_api/onnxruntime_training_c_api.cc#L337-L342).

Not sure what you think is cleaner. The difference there is that the user is required to free up the memory as needed by their application.

Reply:
This is a copy of a function the function that is present in inference API. Is there a way to avoid code duplication?

Reply:
@baijumeswani Here the function is used by the user to create a NamedOnnxValue object in C#. The lifetime is the entire training session. Since InferenceSession uses the allocator implementation for similar methods, I want to align with that implementation. 
See GetTensorRTProviderOptionsAsString, GetCUDAProviderOptionsAsString, SessionGetInputName etc.

Reply:
I understand why it is needed. Can we put this into a common place so both inference and training can use it?